### ISSUE TITLE
Coverage fails with Robolectric tests in AndroidStudio

### ISSUE THREAD

**Issue by Squeazer** [robolectric/robolectric#3023]
### Description
When running robolectric tests with coverage in android studio, those tests will fail wil a `java.lang.VerifyError: Bad return type`. Running tests normally works OK. 
Full log: [robolectric_crash_log.txt](https://github.com/robolectric/robolectric/files/907424/robolectric_crash_log.txt)

### Steps to Reproduce

1. Write a Robolectric test, e.g.:
```java
@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class)
public class DeviceManagerTest {
    @Test
    public void testMe() {
        assertTrue(1 + 1 == 2);
    }
}
```
2. Run the test with coverage
### Robolectric & Android Version

- Robolectric: 3.3.2 (But this happens from 3.1 onward)
- Android:
  - minSdkVersion 19
  - targetSdkVersion 25
  - buildToolsVersion "24.0.3"
  - Android Studio 2.3.1

**Comment by hrskrs** [comment]
same happens with:
minSdkVersion 16
targetSdkVersion 25
buildTools 25.0.1
AndroidStudio 2.4 preview 7

**Comment by zhangzhxb520** [comment]
same happens with:

```
minSdkVersion 15
targetSdkVersion 25
buildToolsVersion "25.0.2"
Android Studio 2.3.2
```

**Comment by Kostik8** [comment]
the same here. It works well without coverage and fails with coverage. 

tested with Robolectric 3.3.2 minSdk = 19, buildTools = 25.0.3
android Studio 2.3.2, 2.4 Preview, 3.0 Preview - all the same.

Works perfectly with robolectric 3.2.1

**Comment by Blubbaz** [comment]
Still the same. 

Coverage doesn't work since 3.2.1.

**Comment by jamesjmtaylor** [comment]
Same here, tests can be run normally, but attempting to run them with code coverage leads to a run-time error with no code coverage report generated.  

Robolectric 3.3.2
Android studio 2.3.3, minSdkVersion 19, compileSdkVersion 25, buildToolsVersion 25.0.2.

**Comment by lijing0283** [comment]
Same issue. 

Robolectric 3.4.2
Android studio 2.3.2, minSdkVersion 21, compileSdkVersion 26, buildToolsVersion 26.0.1, androidSupportVersion 26.0.1

Tests can be run well but coverage fails. (Some of our test cases need to be running on specific API versions. We use config to specify the API versions for some test classes and this might cause the coverage failure. Because after I remove them, coverage will pass.)

However, works perfectly with Robolectric 3.2.1 (Even apply config to specify API versions).

**Comment by kingargyle** [comment]
Which coverage plugin are you using in Android Studio.  It comes with two built in.   IntelliJ's coverage, and Jacoco.   I've seen tests have issues with IntelliJs's coverage plugin, you can manually switch it to use Jacoco on the Coverage tab for the test if it isn't already using it.

**Comment by jamesjmtaylor** [comment]
I don't see any way to switch in the coverage tab: ![screen shot 2017-08-28 at 5 35 48 am](https://user-images.githubusercontent.com/17860994/29773547-d3b2f1ce-8bb2-11e7-8dd4-2ad400594a69.png)

**Comment by kingargyle** [comment]
https://www.jetbrains.com/help/idea/configuring-code-coverage-measurement.html 

While the above link is for IntelliJ, the code coverage options in Android Studio 2.3 were in the same place.

**Comment by jamesjmtaylor** [comment]
Cool beans.  That seemed to do it for me.  I'm including a couple screenshots of the flow to change the coverage runner for the other posters.  I'd ask that you keep the ticket open until at least a couple other people confirm that it works for them as well.

![screen shot 2017-08-28 at 7 10 20 am](https://user-images.githubusercontent.com/17860994/29777385-aa850e82-8bc0-11e7-922e-7d46406c329b.png)
![screen shot 2017-08-28 at 7 10 36 am](https://user-images.githubusercontent.com/17860994/29777391-b15a2ae4-8bc0-11e7-87df-7662ff1cd76c.png)

**Comment by aleksandra-majchrzak** [comment]
In my case, when switch to JaCoCo tests pass, but no coverage is shown (everywhere there is 0%).

**Comment by alt236** [comment]
@alealeksandra-majchrzak try cleaning your build before running the tests

**Comment by aleksandra-majchrzak** [comment]
I have already found out what the problem was. In my project I used both Robolectric and PowerMock. With Robolectric, JaCoCo solves the problem, but PowerMock is incompatible with it. However, the problem occurs only when I try to run all the test classes at the same time, so for now I have to stick to manually firing them one after another.

**Comment by esyd1pi0** [comment]
We have upgraded to robolectric 3.4.2 and running the unit tests is creating an empty (zero bytes) jacoco code coverage .exec file now. Any ideas?

**Comment by kingargyle** [comment]
@esyd1pi0 does this happen in Android Studio or at the command prompt?  Certain versions of gradle and jacoco aren't compatible.  If you are using the latest versions of gradle you need to use the latest version of jacoco.

See: https://github.com/jacoco/jacoco/pull/288#issuecomment-185997232

**Comment by ZeonXX** [comment]
Same issue.  
robolectric 3.4.2
With JaCoCo tests pass normaly.

**Comment by DCRichards** [comment]
Also present with robolectric `3.5.1` and gradle plugin `3.0.0` in latest Android Studio 3

**Comment by goncalossilva** [comment]
And Robolectric 3.6.1 + Gradle plug-in 3.0.1.

**Comment by ignaciotcrespo** [comment]
<details>
<summary>Same here with Robolectric 3.6.1 + Gradle plug-in 3.0.1, expand for stacktrace logs</summary>

	java.lang.VerifyError: Bad return type
	Exception Details:
	  Location:
    	android/content/res/ResourcesImpl.$$robo$$loadComplexColorForCookie(Landroid/content/res/Resources;Landroid/util/TypedValue;ILandroid/content/res/Resources$Theme;)Landroid/content/res/ComplexColor; @565: areturn
	  Reason:
	    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'android/content/res/ComplexColor' (from method signature)
	  Current Frame:
	    bci: @565
	    flags: { }
	    locals: { 'android/content/res/ResourcesImpl', 'android/content/res/Resources', 'android/util/TypedValue', integer, 'android/content/res/Resources$Theme', 'java/lang/Object', 'java/lang/String', 'java/lang/Object', 'android/content/res/XmlResourceParser', 'android/util/AttributeSet', integer, 'java/lang/String' }
    stack: { 'java/lang/Object' }
	  Bytecode:
	    0x0000000: 1266 b800 6c3a 0519 0511 03c8 b800 702c
	    0x0000010: b401 79c7 002a 1905 1103 c9b8 0070 bb03
	    0x0000020: a159 bb00 eb59 b700 ec13 03a3 b600 f22c
	    0x0000030: b402 7fb6 014f b600 fcb7 03a4 bf19 0511
	    0x0000040: 03cd b800 702c b401 79b9 017a 0100 3a06
	    0x0000050: 1905 1103 deb8 0070 013a 0719 0511 03e0
	    0x0000060: b800 7014 019c 1906 b801 a519 0511 03e1
	    0x0000070: b800 7019 0613 0312 b603 1699 0154 1905
	    0x0000080: 1103 e3b8 0070 2a19 061d 2cb4 0177 1303
	    0x0000090: a6b6 031a 3a08 1905 1103 e6b8 0070 1908
	    0x00000a0: b803 ac3a 0919 0511 03e8 b800 7019 08b9
	    0x00000b0: 03af 0100 5936 0a05 9f00 0c15 0a04 9f00
	    0x00000c0: 06a7 ffe4 1905 1103 ecb8 0070 150a 059f
	    0x00000d0: 0016 1905 1103 edb8 0070 bb03 3159 1303
	    0x00000e0: b1b7 03b2 bf19 0511 03f0 b800 7019 08b9
	    0x00000f0: 03b5 0100 3a0b 1905 1103 f1b8 0070 190b
	    0x0000100: 1303 b7b6 0162 9900 1a19 0511 03f2 b800
	    0x0000110: 702b 1908 1909 1904 b803 bd3a 07a7 002a
	    0x0000120: 1905 1103 f3b8 0070 190b 1303 bfb6 0162
	    0x0000130: 9900 1719 0511 03f4 b800 702b 1908 1909
	    0x0000140: 1904 b803 c23a 0719 0511 03f6 b800 7019
	    0x0000150: 08b9 0321 0100 1905 1103 feb8 0070 a700
	    0x0000160: bf19 0511 03f7 b800 703a 0819 0511 03f8
	    0x0000170: b800 7014 019c b802 2f19 0511 03f9 b800
	    0x0000180: 70bb 0011 59bb 00eb 59b7 00ec 1301 80b6
	    0x0000190: 00f2 1906 b600 f213 03c4 b600 f21d 1905
	    0x00001a0: 1103 fbb8 0070 b800 f8b6 00f2 b600 fcb7
	    0x00001b0: 00ff 3a09 1905 1103 fcb8 0070 1909 1908
	    0x00001c0: b601 9457 1905 1103 fdb8 0070 1909 bf19
	    0x00001d0: 0511 0400 b800 7014 019c b802 2f19 0511
	    0x00001e0: 0401 b800 70bb 0011 59bb 00eb 59b7 00ec
	    0x00001f0: 1301 80b6 00f2 1906 b600 f213 0182 b600
	    0x0000200: f21d 1905 1104 03b8 0070 b800 f8b6 00f2
	    0x0000210: 1303 8db6 00f2 b600 fcb7 00ff bf19 0511
	    0x0000220: 0405 b800 7014 019c b802 2f19 0511 0407
	    0x0000230: b800 7019 07b0                         
	  Exception Handler Table:
	    bci [126, 342] => handler: 353
	  Stackmap Table:
	    append_frame(@61,Object[#4])
	    full_frame(@165,	{Object[#2],Object[#19],Object[#372],Integer,Object[#22],Object[#4],Object[#145],Null,Object[#800],Object[#969]},{})
	    append_frame(@196,Integer)
	    same_frame(@229)
	    append_frame(@288,Object[#145])
	    full_frame(@327,	{Object[#2],Object[#19],Object[#372],Integer,Object[#22],Object[#4],Object[#145],Object[#4],Object[#800],Object[#969],Integer,Object[#145]},{})
	    full_frame(@353,	{Object[#2],Object[#19],Object[#372],Integer,Object[#22],Object[#4],Object[#145],Object[#4]},	{Object[#279]})
    full_frame(@463,	{Object[#2],Object[#19],Object[#372],Integer,Object[#22],Object[#4],Object[#145],Null},{})
    full_frame(@541,	{Object[#2],Object[#19],Object[#372],Integer,Object[#22],Object[#4],Object[#145],Object[#4],Object[#800],Object[#969],Integer,Object[#145]},{})


	at android.content.res.Resources.__constructor__(Resources.java:219)
	at android.content.res.Resources.<init>(Resources.java)
	at org.robolectric.shadows.ShadowResources.getSystem(ShadowResources.java:92)
	at android.content.res.Resources.getSystem(Resources.java)
	at org.robolectric.android.internal.ParallelUniverse.setUpApplicationState(ParallelUniverse.java:116)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:328)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:241)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:123)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:77)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)

</details>

**Comment by Rowandjj** [comment]
@ignaciotcrespo that works for me too...

**Comment by KaidoNinja** [comment]
@ignaciotcrespo This only works locally. If you use, for example, Bamboo + SonarQube, then it still fails. We need an actual fix not a local workaround.

**Comment by ignaciotcrespo** [comment]
@KaidoNinja totally agree with you! I hope the robolectric team find a fix for this soon.

**Comment by k4biri** [comment]
Also present with robolectric 3.7.1 and gradle plugin 3.1.0 in latest Android Studio 3.1 .

**Comment by soldier4443** [comment]
Present with Robolectric 3.8 + gradle 3.1.0 .. Well, I'm gonna use @ignaciotcrespo's solution for now!

**Comment by kyawsanoo** [comment]
In my case(Roboelectric 3.8, android tool version 27.0.3 and gradle plugin 3.1.3), when I run with specific test with code coverage, it's OK but when I run all tests with coverage, all tests fail and get error as follow:

java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    android/content/res/ResourcesImpl.$$robo$$loadComplexColorForCookie(Landroid/content/res/Resources;Landroid/util/TypedValue;ILandroid/content/res/Resources$Theme;)Landroid/content/res/ComplexColor; @565: areturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'android/content/res/ComplexColor' (from method signature)
  Current Frame:
    bci: @565
    flags: { }
    locals: { 'android/content/res/ResourcesImpl', 'android/content/res/Resources', 'android/util/TypedValue', integer, 'android/content/res/Resources$Theme', 'java/lang/Object', 'java/lang/String', 'java/lang/Object', 'android/content/res/XmlResourceParser', 'android/util/AttributeSet', integer, 'java/lang/String' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0x0000000: 127c b800 823a 0519 0511 041b b800 862c
    0x0000010: b401 a9c7 002a 1905 1104 1cb8 0086 bb04
    0x0000020: 2859 bb01 1b59 b701 1c13 042a b601 222c
    0x0000030: b402 b3b6 017f b601 2cb7 042b bf19 0511
    0x0000040: 0420 b800 862c b401 a9b9 01aa 0100 3a06
    0x0000050: 1905 1104 31b8 0086 013a 0719 0511 0433
    0x0000060: b800 8614 01d0 1906 b801 d919 0511 0434
    0x0000070: b800 8619 0613 0371 b603 7599 0154 1905
    0x0000080: 1104 36b8 0086 2a19 061d 2cb4 01a7 1304
    0x0000090: 2db6 0379 3a08 1905 1104 39b8 0086 1908
    0x00000a0: b804 333a 0919 0511 043b b800 8619 08b9
    0x00000b0: 0436 0100 5936 0a05 9f00 0c15 0a04 9f00
    0x00000c0: 06a7 ffe4 1905 1104 3fb8 0086 150a 059f
    0x00000d0: 0016 1905 1104 40b8 0086 bb03 b959 1304
    0x00000e0: 38b7 0439 bf19 0511 0443 b800 8619 08b9
    0x00000f0: 043c 0100 3a0b 1905 1104 44b8 0086 190b
    0x0000100: 1304 3eb6 0192 9900 1a19 0511 0445 b800
    0x0000110: 862b 1908 1909 1904 b804 443a 07a7 002a
    0x0000120: 1905 1104 46b8 0086 190b 1304 46b6 0192
    0x0000130: 9900 1719 0511 0447 b800 862b 1908 1909
    0x0000140: 1904 b804 493a 0719 0511 0449 b800 8619
    0x0000150: 08b9 0380 0100 1905 1104 51b8 0086 a700
    0x0000160: bf19 0511 044a b800 863a 0819 0511 044b
    0x0000170: b800 8614 01d0 b802 6319 0511 044c b800
    0x0000180: 86bb 0016 59bb 011b 59b7 011c 1301 b0b6
    0x0000190: 0122 1906 b601 2213 044b b601 221d 1905
    0x00001a0: 1104 4eb8 0086 b801 28b6 0122 b601 2cb7
    0x00001b0: 012f 3a09 1905 1104 4fb8 0086 1909 1908
    0x00001c0: b601 c857 1905 1104 50b8 0086 1909 bf19
    0x00001d0: 0511 0453 b800 8614 01d0 b802 6319 0511
    0x00001e0: 0454 b800 86bb 0016 59bb 011b 59b7 011c
    0x00001f0: 1301 b0b6 0122 1906 b601 2213 01c4 b601
    0x0000200: 221d 1905 1104 56b8 0086 b801 28b6 0122
    0x0000210: 1304 15b6 0122 b601 2cb7 012f bf19 0511
    0x0000220: 0458 b800 8614 01d0 b802 6319 0511 045a
    0x0000230: b800 8619 07b0                         
  Exception Handler Table:
    bci [126, 342] => handler: 353
  Stackmap Table:
    append_frame(@61,Object[#4])
    full_frame(@165,{Object[#2],Object[#24],Object[#420],Integer,Object[#27],Object[#4],Object[#194],Null,Object[#895],Object[#1104]},{})
    append_frame(@196,Integer)
    same_frame(@229)
    append_frame(@288,Object[#194])
    full_frame(@327,{Object[#2],Object[#24],Object[#420],Integer,Object[#27],Object[#4],Object[#194],Object[#4],Object[#895],Object[#1104],Integer,Object[#194]},{})
    full_frame(@353,{Object[#2],Object[#24],Object[#420],Integer,Object[#27],Object[#4],Object[#194],Object[#4]},{Object[#327]})
    full_frame(@463,{Object[#2],Object[#24],Object[#420],Integer,Object[#27],Object[#4],Object[#194],Null},{})
    full_frame(@541,{Object[#2],Object[#24],Object[#420],Integer,Object[#27],Object[#4],Object[#194],Object[#4],Object[#895],Object[#1104],Integer,Object[#194]},{})


	at android.content.res.Resources.__constructor__(Resources.java:220)
	at android.content.res.Resources.<init>(Resources.java)
	at org.robolectric.shadows.ShadowResources.getSystem(ShadowResources.java:92)
	at android.content.res.Resources.getSystem(Resources.java)
	at org.robolectric.android.internal.ParallelUniverse.setUpApplicationState(ParallelUniverse.java:161)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:335)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:245)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)


java.lang.IllegalStateException: this method should only be called by Robolectric

	at org.robolectric.shadows.ShadowDisplayManager.configureDefaultDisplay(ShadowDisplayManager.java:43)
	at org.robolectric.android.Bootstrap.setUpDisplay(Bootstrap.java:19)
	at org.robolectric.android.internal.ParallelUniverse.setUpApplicationState(ParallelUniverse.java:159)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:335)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:245)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)

**Comment by kyaw-san-oo** [comment]
In Android Studio to get rid of this error I have to append the -ea -noverify to the VMOptions in the EditConfigurations of RobolectricTests. Intellij Idea Code Coverage now success for all reboelectric tests in my case.

**Comment by abangfadli** [comment]
I had this issue yesterday and in my case it's just simply update the version of robolectric to `1.6.6` as mentioned in [here](https://github.com/robolectric/robolectric/wiki/Using-PowerMock)

**Comment by peterfortuin** [comment]
I'm running Robolectric 4.1 and I still have this problem. 
The -noverify works for Android Studio, but it doesn't help me when I run the test with Gradle in a CI environment.

**Comment by peterfortuin** [comment]
> @peterfortuin try this, it works for me: [card-io/card.io-Android-SDK#63 (comment)](https://github.com/card-io/card.io-Android-SDK/issues/63#issuecomment-239144782)

Thanks for the tip, but somehow that doesn't seem to work for me. Not sure why.

**Comment by peterfortuin** [comment]
Thanks a lot. For me it was the following lines that made it working:

`
tasks.withType(Test) {
    jacoco.includeNoLocationClasses = true
}
`

**Comment by meness** [comment]
> Same as @goncalossilva with Robolectric 3.6.1 + Gradle plug-in 3.0.1, expand for stacktrace logs
> #### UPDATE: In my case I need to use the AS internal coverage runner instead of Jacoco, it works using a workaround, adding `-noverify` to VM options in Android Studio
> ![image](https://user-images.githubusercontent.com/6496116/34828090-23092a90-f6bb-11e7-8d64-7d3f6f2aef06.png)

I had the issue and the above reply fixed it. Choosing Jacoco fixes as well but I prefer the `-noverify` argument method.

**Comment by roccadev** [comment]
Having this issue, but none of the proposed workarounds was useful. With JaCoCo, unit tests run succesfully but the report panel is stuck on "loading" and it doesn't show any number. On the other hand, the `-noverify` option with the default Intellij runner doesn't change anything.

Android Studio 3.3.2
Robolectric 4.2
Kotlin code

```
testOptions.unitTests.includeAndroidResources = true
```

Excerpt from the report:
```
java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    android/content/res/ResourcesImpl.$$robo$$android_content_res_ResourcesImpl$loadComplexColorForCookie(Landroid/content/res/Resources;Landroid/util/TypedValue;ILandroid/content/res/Resources$Theme;)Landroid/content/res/ComplexColor; @951: areturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'android/content/res/ComplexColor' (from method signature)

at android.content.res.Resources.__constructor__(Resources.java:228)
	at android.content.res.Resources.<init>(Resources.java)
	at android.content.pm.PackageParser.parseBaseApk(PackageParser.java:1334)
	at android.content.pm.PackageParser.parseMonolithicPackage(PackageParser.java:1299)
	at android.content.pm.PackageParser.parsePackage(PackageParser.java:1022)
	at android.content.pm.PackageParser.parsePackage(PackageParser.java:1042)
	at org.robolectric.shadows.ShadowPackageParser.callParsePackage(ShadowPackageParser.java:41)
	at org.robolectric.android.internal.AndroidEnvironment.loadAppPackage_measured(AndroidEnvironment.java:291)
	at org.robolectric.android.internal.AndroidEnvironment.lambda$loadAppPackage$1(AndroidEnvironment.java:261)
	at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:53)
	at org.robolectric.android.internal.AndroidEnvironment.loadAppPackage(AndroidEnvironment.java:259)
	at org.robolectric.android.internal.AndroidEnvironment.installAndCreateApplication(AndroidEnvironment.java:158)
	at org.robolectric.android.internal.AndroidEnvironment.setUpApplicationState(AndroidEnvironment.java:149)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:298)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:247)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
```

**Comment by uOOOO** [comment]
Same here,

Android Studio 3.4, Robolectric 4.2 and Kotlin

```
java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    android/content/res/ResourcesImpl.$$robo$$android_content_res_ResourcesImpl$loadComplexColorForCookie(Landroid/content/res/Resources;Landroid/util/TypedValue;ILandroid/content/res/Resources$Theme;)Landroid/content/res/ComplexColor; @565: areturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'android/content/res/ComplexColor' (from method signature)
  Current Frame:
    bci: @565
    flags: { }
    locals: { 'android/content/res/ResourcesImpl', 'android/content/res/Resources', 'android/util/TypedValue', integer, 'android/content/res/Resources$Theme', 'java/lang/Object', 'java/lang/String', 'java/lang/Object', 'android/content/res/XmlResourceParser', 'android/util/AttributeSet', integer, 'java/lang/String' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0x0000000: 128a b800 903a 0519 0511 0437 b800 942c
    0x0000010: b401 b1c7 002a 1905 1104 38b8 0094 bb04
    0x0000020: 4359 bb01 2359 b701 2413 0445 b601 2a2c
```

**Comment by tschellenbach** [comment]
this is still broken

**Comment by 47even** [comment]
Test coverage with JaCoCo runner works fine with robolectric 4.3.1

**Comment by neiljaywarner** [comment]
bump. This drives me crazy and has been going for a year.  I'm not able to get it from within android studio with noverify, with jacoco or with intellij.  It would really be nice to have this for when we are asked.

**Comment by hoisie** [comment]
@neiljaywarner a lot of other folks in this bug thread appear to have coverage working. If it doesn't work for you, maybe you are running into a separate issue? If this is the case, please create a new issue and follow  https://github.com/robolectric/robolectric/wiki/How-to-Report-Issues

Also, in the spirit of transparency, due to many other priorities I will not have time to investigate this. If this is crucial for you I would recommend that you investigate the root cause of this issue.

**Comment by carstenhag** [comment]
Works with `-ea -noverify` and IntelliJ runner.

But imo this should not be necessary. Maybe this is caused by the IntelliJ runner? Is there an issue upstream at Jetbrains/IntelliJ about this?

**Comment by utzcoz** [comment]
[Google I/O 2021 What's new in Android testing tool](https://www.youtube.com/watch?v=juEkViDyzF8) says that Android Studio Arctic Fox has delegating tests to new Android Gradle Test Runner, and deprecate the usage of Intellij Test Runner. There are two screenshots about it:

![Old test sequence with Gradle Test Runner and Intellij Test Runner](https://user-images.githubusercontent.com/43091780/119221754-5d63c080-bb23-11eb-907f-d1c211a04b54.png)
![New test sequence with Android Gradle Test Runner](https://user-images.githubusercontent.com/43091780/119221756-605eb100-bb23-11eb-871e-db68dfe8c80b.png)

Maybe it is useful to help to fix this problem or other potential Intellij Test Runner related problems. cc @hoisie .

**Comment by charlesng** [comment]
> @peterfortuin
> 
> Hi Peter I had the same issue what I am using is following gradle task (this on is working, please check on the mainSrc and filetree it's for kotlin):
> 
> ```
> task jacocoTestReport(type: JacocoReport, dependsOn: ['testDebugUnitTest', 'createDebugCoverageReport']) {
> 
>     reports {
>         xml.enabled = true
>         html.enabled = true
>     }
> 
>     def fileFilter = [ '**/R.class', '**/R$*.class', '**/BuildConfig.*', '**/Manifest*.*', '**/*Test*.*', 'android/**/*.*' ]
>     def debugTree = fileTree(dir: "$project.buildDir/tmp/kotlin-classes/debug", excludes: fileFilter)
>     def mainSrc = "$project.projectDir/src/main/kotlin"
> 
>     sourceDirectories = files([mainSrc])
>     classDirectories = files([debugTree])
>     executionData = fileTree(dir: project.buildDir, includes: [
>             'jacoco/testDebugUnitTest.exec', 'outputs/code-coverage/connected/*coverage.ec'
>     ])
> }
> ```
> 
> next step I am using Robolectric 4.1 and jacoco 0.8.2 and following configs inside app => build.gradle
> 
> ```
> jacoco {
>     toolVersion = '0.8.2'
> }
> 
> tasks.withType(Test) {
>     jacoco.includeNoLocationClasses = true
> }
> 
>     testOptions {
> 
>         animationsDisabled true
> 
>         unitTests {
>             includeAndroidResources = true
>         }
>         unitTests.all {
>             systemProperty 'robolectric.enabledSdks', '28'
> 			jvmArgs '-noverify'
>         }
>     }
> 
>     buildTypes {
>         debug {
>             testCoverageEnabled true
>         }
>         release {
>             minifyEnabled false
>             testCoverageEnabled true
>             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
>         }
>     }
> ```
> 
> and finally for testing I run ./gradlew build jacocoTestReport
> I found the solution right now, so I don't know which parameter was it, but jacocoTestReport is one part of the solution

This solution works for Robolectric 4.6 and Android Studio 4.3

**Comment by alwa** [comment]
I still see this issue in Robolectric 4.6.1 and Android Studio 2020.3.1 (Patch 1):

```
Bad return type
Exception Details:
  Location:
    android/content/res/ResourcesImpl.$$robo$$android_content_res_ResourcesImpl$loadComplexColorForCookie(Landroid/content/res/Resources;Landroid/util/TypedValue;ILandroid/content/res/Resources$Theme;)Landroid/content/res/ComplexColor; @803: areturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'android/content/res/ComplexColor' (from method signature)
  Current Frame:
    bci: @803
    flags: { }
    locals: { 'android/content/res/ResourcesImpl', 'android/content/res/Resources', 'android/util/TypedValue', integer, 'android/content/res/Resources$Theme', '[Z', 'java/lang/Object', 'java/lang/String', 'java/lang/Object', 'android/content/res/XmlResourceParser', 'android/util/AttributeSet', integer, 'java/lang/String' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0000000: b805 da3a 0512 7cb8 0082 3a06 1905 1102
    0000010: 0604 5419 0611 041b b800 862c b401 a9c7
    0000020: 0038 1905 1102 0704 5419 0611 041c b800
    0000030: 86bb 042a 59bb 011b 59b7 011c 1304 2cb6
    0000040: 0122 2cb4 02b5 b601 7fb6 012c b704 2d19
    0000050: 0511 0208 0454 bf19 0611 0420 b800 862c
    0000060: b401 a9b9 01aa 0100 3a07 1905 1102 0904
    0000070: 5419 0611 0431 b800 8601 3a08 1905 1102
    0000080: 0a04 5419 0611 0433 b800 8614 01d2 1907
etc
```

My test class is annoted as follows:
```
@Config(sdk = [Build.VERSION_CODES.N])
@RunWith(AndroidJUnit4::class)
@LooperMode(LooperMode.Mode.PAUSED)
```

build.gradle.kts:
```
testOptions {
        unitTests.isReturnDefaultValues = true
        unitTests.isIncludeAndroidResources = true
...
}
```

**Comment by anidhamal** [comment]
@alwa did you find any workaround? I'm having same issue.

**Comment by alwa** [comment]
> @alwa did you find any workaround? I'm having same issue.

No, unfortunately not.

**Comment by hoisie** [comment]
Looks like jacoco has the ability to run its instrumentation offline: https://www.jacoco.org/jacoco/trunk/doc/offline.html

@alwa I would recommend trying to run jacoco instrumentation on the preinstrumented Robolectric jar and look at the bytecode of loadComplexColorForCookie to see the difference to the regular one.

**Comment by nicolas71640** [comment]
On Android Studio Arctic Fox, Robolectric doesn't allow the coverage anymore (Bad return type error). The -noverify flag doesn't work anymore. And I couldn't use Jacoco to have the coverage inside Android Studio (use Jacoco to generate the html report in the CI, this work). 
The solution => downgrade AS to 4.2.2

**Comment by Abdul96881** [comment]
cant even access android resources also

**Comment by utzcoz** [comment]
One update maybe related to this problem: https://androidstudio.googleblog.com/2022/02/android-studio-bumblebee-202111-patch-2.html Android Studio BumbleBee 202111 Patch2 released with some test coverage related fixes.

1. [Issue #213534628](https://buganizer.corp.google.com/issues/213534628): Android Gradle Plugin 7.1.0-rc01 fails to perform ASM bytecode transformation during unit tests
2. [Issue #210148618](https://buganizer.corp.google.com/issues/210148618): java.lang.NoSuchMethodError: No static method $jacocoInit
3. [Issue #210485236](https://buganizer.corp.google.com/issues/210485236): Android test coverage broken due to Mockito error

Maybe it deserves a try.

**Comment by hoisie** [comment]
Is there a way to configure the Jacoco agent to exclude instrumenting certain packages? For instance, there is really no benefit for Jacoco to instrument the `android.*` classes, because there is no purpose to get code coverage for these. Although this wouldn't fix the issue of Robolectric instrumentation clashing with Jacoco, it'll at least work around the problem, and likely even improve performance when running coverage.

**Comment by hoisie** [comment]
Looks like if there is a resource file `jacoco-agent.properties`, you can potentially configure the agent to avoid instrumenting the Android classes:

https://groups.google.com/g/jacoco/c/vx0g_6TKY8Q
https://github.com/Godin/jacoco-experiments/blob/android/android/app.tests/src/main/resources/jacoco-agent.properties
https://www.eclemma.org/jacoco/trunk/doc/agent.html

I am not sure if this supports package wildcards like `android.*`

**Comment by ZSmallX** [comment]
@hoisie I had investigated a similar problem(coverage 0% when Jacoco works with PowerMock) before and make a HACK work-around working in my work. I think I can share the solution and code here.

Investigations:  
The main reason is that both Jacoco,  most of Mock frameworks and Robolectric modified the class at runtime, and the instrumentation order is sensitive to all those tools. So generally it's hard to make all tools working well together.

References:  
- Coverage of runtime modified classes https://github.com/jacoco/jacoco/issues/51
- Use ByteBuddy to instrument classes instead Javassist https://github.com/powermock/powermock/issues/727
- Code Coverage with JaCoCo https://github.com/powermock/powermock/wiki/Code-coverage-with-JaCoCo

But luckly JaCoCo supports instrumentation offline. and unluckly JaCoCo Gradle Plugin only supports JaCoCo on-the-fly.

See: Add offline instrumentation support to Gradle Jacoco plugin https://github.com/gradle/gradle/issues/2429

Thus my solution is to make it possible to support JaCoCo offline instrumentation in Gradle and make the runtime classpath right with JaCoCo on-the-fly disabled.

![image](https://user-images.githubusercontent.com/14019790/157483572-28697ca8-ab6e-4b5b-b859-cc61dbd00897.png)

Core code snippets:  

```
apply plugin: 'jacoco'

jacoco {
    toolVersion = "0.8.3"
}

configurations {
    jacocoAnt
    jacocoRuntime
}

afterEvaluate { project ->

androidVariants = android.libraryVariants

        androidVariants.each { androidVariant ->

        javaCompile = androidVariant.javaCompileProvider.get()

        // Define instrument task
        task "${jacocoInstrumentTaskName}"(dependsOn: [compileSourceTaskName]) {
                group = "Jacoco"
                description = "Generate instrumented classes on the ${variantName.capitalize()} build."
                println "[${group}]:Defining task :${project.name}:${name}"
                doLast {
                    println "[${group}]:Generating instrumented classes on the ${variantName.capitalize()} build."
                    ant.taskdef(name: 'instrument',
                            classname: 'org.jacoco.ant.InstrumentTask',
                            classpath: configurations.jacocoAnt.asPath)
                    ant.instrument(destdir: jacocoInstrumentOutputDir) {
                        fileset(
                                dir: jacocoInstrumentInputDir,
                                excludes: [
                                        '**/R.class',
                                        '**/R$*.class',
                                        '**/*$ViewInjector*.*',
                                        '**/BuildConfig.*',
                                        '**/Manifest*.*'
                                ]
                        )
                    }
                    if (isKotlinContained) {
                        ant.instrument(destdir: jacocoInstrumentOutputDir) {
                            fileset(
                                    dir: kotlinInstrumentInputDir,
                                    excludes: [
                                            '**/R.class',
                                            '**/R$*.class',
                                            '**/*$ViewInjector*.*',
                                            '**/BuildConfig.*',
                                            '**/Manifest*.*'
                                    ]
                            )
                        }
                    }
                }
            }

def unitTestTaskName = "test" + androidVariant.unitTestVariant.name.capitalize()
            def executionDataDir = "${buildDir.path}/jacoco/${unitTestTaskName}.exec"

            // ...

            // Put instrumented classes and JacocoRuntime to the beginning of classpath
            tasks[unitTestTaskName].doFirst {
                jacoco {
                    // Disabled on-the-fly
                    // https://github.com/jacoco/jacoco/issues/51
                    // https://github.com/powermock/powermock/issues/727
                    // https://github.com/powermock/powermock/wiki/Code-coverage-with-JaCoCo
                    enabled = false
                }

                println "[Jacoco]:Modifying args of ${unitTestTaskName}"
                systemProperty 'jacoco-agent.destfile', executionDataDir
                classpath = files(jacocoInstrumentOutputDir) + classpath + configurations.jacocoRuntime
                println "[Jacoco]:Running ${name}"
                for (def path in classpath.getAsPath().split(":")) {
                    if (path.contains("jacoco")) {
                        println "[Jacoco]: jacoco related classpath of project: ${path}"
                    }

                    if (!path.startsWith(buildDir.path)) {
                        continue
                    }
                    println "[Jacoco]:classpath of project: ${path}"
                }
            }

            def reportTaskName = "create${variantName.capitalize()}CodeCoverage"

            def sourceFiles = []
            def sourceFilePaths = []
            androidVariant.sourceSets.each { sourceSet ->
                sourceFiles.addAll(sourceSet.javaDirectories)
                sourceSet.javaDirectories.each { javaDirectory ->
                    if (javaDirectory.exists()) {
                        sourceFilePaths.add(javaDirectory.canonicalPath)
                    }
                }
            }

            // Define report task
            task "${reportTaskName}"(type: JacocoReport, dependsOn: [jacocoInstrumentTaskName, unitTestTaskName]) {
                group = "Jacoco"
                description = "Generate Jacoco coverage reports on the ${variantName.capitalize()} build."
                println "[${group}]:Defining task :${project.name}:${name}"

                def excludeFiles = [
                        '**/R.class',
                        '**/R$*.class',
                        '**/*$ViewInjector*.*',
                        '**/*$ViewBinder*.*',
                        '**/BuildConfig.*',
                        '**/Manifest*.*'
                ]

                def extraExecFilePathsStr = project.getProperties().get('exec_path')
                def extraExcludeFilesStr = project.getProperties().get('exclude_files')

                if (extraExcludeFilesStr != null) {
                    def extraExcludeFiles = extraExcludeFilesStr.split(',')
                    excludeFiles.addAll(extraExcludeFiles)
                }

                def execFilePaths = [executionDataDir]

                if (extraExecFilePathsStr != null) {
                    def extraExecFiles = extraExecFilePathsStr.split(',')
                    execFilePaths.add(extraExecFiles)
                }


                def javaClassDirectories = fileTree(
                        dir: jacocoInstrumentInputDir,
                        excludes: excludeFiles
                )

                def kotlinClassDirectories = fileTree(
                        dir: kotlinInstrumentInputDir,
                        excludes: excludeFiles
                )

                def gradleVersion = gradle.gradleVersion

                println "Running gradle version: $gradleVersion"


                if (gradleVersion.startsWith('4.')) {
                    println "Use compatible set method."

                    setExecutionData(files(execFilePaths))
                    if (isKotlinContained) {
                        setClassDirectories(javaClassDirectories, kotlinClassDirectories)
                    } else {
                        setClassDirectories(javaClassDirectories)
                    }
                    setSourceDirectories(sourceFiles)
                } else {
                    getExecutionData().setFrom(files(execFilePaths))
                    if (isKotlinContained) {
                        getClassDirectories().setFrom(javaClassDirectories, kotlinClassDirectories)
                    } else {
                        getClassDirectories().setFrom(javaClassDirectories)
                    }
                    getSourceDirectories().setFrom(sourceFiles)
                }

                reports {
                    xml.enabled = true
                    html.enabled = true
                }

                doLast {
                    def reportFile = "${buildDir.path}/reports/jacoco/${reportTaskName}/html/index.html"
                    println "[Jacoco]:Report file generated at:${reportFile}"
                }
            }
        }
    }
```

I can do some verification on it and make the implementation code public later.

**Comment by ZSmallX** [comment]
I added a JaCoCo integration test module and JaCoCo offline instrumentation script in #7114 and run the tests in Android Studio BumbleBee Patch 2.

Here's the verification results of JaCoCo offline instrumentation:  
No Robolectric exception raised but no coverage in IDE as well, but JaCoCo execution data is available and it can be imported into IDE and get the right code coverage.

I think that JaCoCo offline instrumentation is a suitable way to solve this problem(conflicts in runtime class loaders and class modification), but IDE prefers the java agent solution to collect the coverage data and get benifits from it.

See:  https://www.jacoco.org/jacoco/trunk/doc/offline.html

> One of the main benefits of JaCoCo is the Java agent, which instruments classes on-the-fly. This simplifies code coverage analysis a lot as no pre-instrumentation and classpath tweaking is required. However, there can be situations where on-the-fly instrumentation is not suitable, for example:
> - Runtime environments that do not support Java agents.
> - Deployments where it is not possible to configure JVM options.
> - Bytecode needs to be converted for another VM like the Android Dalvik VM.
> - **Conflicts with other agents that do dynamic classfile transformation.**

Thus, this solution **CAN** solves the conflicts between JaCoCo and any frameworks based on javaagent and class modification ad runtime. And make the CI automation work by parsing JaCoCo execution data on JaCoCo report tool.

And **CAN NOT** solves the way IDE collects the coverage data using IDEA coverage java agent and JaCoCo agent.

![image](https://user-images.githubusercontent.com/14019790/157581171-32b6e727-55a9-4472-84f1-ea72aa6c1808.png)

**Comment by rgautam1602** [comment]
I am also facing this issue with android studio chipmunk,could someone please let me know how to resolve

**Comment by hoisie** [comment]
Can someone provide clear repro instructions for this `Bad return type error`? Does it happen on any trivial Robolectric test?

**Comment by Shrujeesh** [comment]
I am using Android Studio Chipmunk
Robo electric version 4.2.1


huh? java 11, need help!
java.lang.RuntimeException: huh? java 11, need help!
	at org.robolectric.util.reflector.UnsafeAccess$Danger11Plus.throwException(UnsafeAccess.java:92)
	at org.robolectric.util.reflector.UnsafeAccess.throwException(UnsafeAccess.java:30)
	at org.robolectric.internal.bytecode.Sandbox.runOnMainThread(Sandbox.java:102)
	at org.robolectric.internal.bytecode.Sandbox.runOnMainThread(Sandbox.java:88)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:226)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:100)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy5.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.RuntimeException: huh? java 11, need help!
	at org.robolectric.util.reflector.UnsafeAccess$Danger11Plus.throwException(UnsafeAccess.java:92)
	at org.robolectric.util.reflector.UnsafeAccess.throwException(UnsafeAccess.java:30)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:261)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more
Caused by: java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    android/content/res/ResourcesImpl.$$robo$$android_content_res_ResourcesImpl$loadComplexColorForCookie(Landroid/content/res/Resources;Landroid/util/TypedValue;ILandroid/content/res/Resources$Theme;)Landroid/content/res/ComplexColor; @827: areturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'android/content/res/ComplexColor' (from method signature)
  Current Frame:
    bci: @827
    flags: { }
    locals: { 'android/content/res/ResourcesImpl', 'android/content/res/Resources', 'android/util/TypedValue', integer, 'android/content/res/Resources$Theme', '[Z', '[I', 'java/lang/String', 'java/lang/Object', 'android/content/res/XmlResourceParser', 'android/util/AttributeSet', integer, 'java/lang/String' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0000000: b805 343a 05b8 0068 b200 6b3a 0619 0611
    0000010: 03c8 5c2e 0460 4f2c b401 74c7 003a 1905
    0000020: 1101 6204 5419 0611 03c9 5c2e 0460 4fbb
    0000030: 039e 59bb 00e6 59b7 00e7 1303 a0b6 00ed
    0000040: 2cb4 027c b601 4ab6 00f7 b703 a119 0511
    0000050: 0163 0454 bf19 0611 03cd 5c2e 0460 4f2c
    0000060: b401 74b9 0175 0100 3a07 1906 1103 de5c
    0000070: 2e04 604f 013a 0819 0511 0164 0454 1906
    0000080: 1103 e05c 2e04 604f 1401 9919 07b8 01a2
    0000090: 1905 1101 6504 5419 0611 03e1 5c2e 0460
    00000a0: 4f19 0713 030f b603 1399 0206 1905 1101
    00000b0: 6604 5419 0611 03e3 5c2e 0460 4f2a 1907
    00000c0: 1d2c b401 7213 03a3 b603 173a 0919 0511
    00000d0: 0167 0454 1906 1103 e65c 2e04 604f 1909
    00000e0: b803 a93a 0a19 0511 0168 0454 1906 1103
    00000f0: e85c 2e04 604f 1909 b903 ac01 0059 360b
    0000100: 05a0 000d 1905 1101 6904 54a7 001d 150b
    0000110: 04a0 000d 1905 1101 6a04 54a7 000d 1905
    0000120: 1101 6b04 54a7 ffc7 1906 1103 ec5c 2e04
    0000130: 604f 150b 059f 0026 1905 1101 6c04 5419
    0000140: 0611 03ed 5c2e 0460 4fbb 032e 5913 03ae
    0000150: b703 af19 0511 016d 0454 bf19 0611 03f0
    0000160: 5c2e 0460 4f19 09b9 03b2 0100 3a0c 1905
    0000170: 1101 6e04 5419 0611 03f1 5c2e 0460 4f19
    0000180: 0c13 03b4 b601 5d99 002a 1905 1101 6f04
    0000190: 5419 0611 03f2 5c2e 0460 4f2b 1909 190a
    00001a0: 1904 b803 ba3a 0819 0511 0170 0454 a700
    00001b0: 4619 0611 03f3 5c2e 0460 4f19 0c13 03bc
    00001c0: b601 5d9a 000d 1905 1101 7104 54a7 0027
    00001d0: 1905 1101 7204 5419 0611 03f4 5c2e 0460
    00001e0: 4f2b 1909 190a 1904 b803 bf3a 0819 0511
    00001f0: 0173 0454 1906 1103 f65c 2e04 604f 1909
    0000200: b903 1e01 0019 0611 03fe 5c2e 0460 4fa7
    0000210: 0109 1906 1103 f75c 2e04 604f 3a09 1905
    0000220: 1101 7404 5419 0611 03f8 5c2e 0460 4f14
    0000230: 0199 b802 2c19 0511 0175 0454 1906 1103
    0000240: f95c 2e04 604f bb00 1159 bb00 e659 b700
    0000250: e713 017b b600 ed19 07b6 00ed 1303 c1b6
    0000260: 00ed 1d19 0511 0176 0454 1906 1103 fb5c
    0000270: 2e04 604f b800 f3b6 00ed b600 f7b7 00fa
    0000280: 3a0a 1905 1101 7704 5419 0611 03fc 5c2e
    0000290: 0460 4f19 0a19 09b6 018f 5719 0611 03fd
    00002a0: 5c2e 0460 4f19 0a19 0511 0178 0454 bf19
    00002b0: 0611 0400 5c2e 0460 4f14 0199 b802 2c19
    00002c0: 0511 0179 0454 1906 1104 015c 2e04 604f
    00002d0: bb00 1159 bb00 e659 b700 e713 017b b600
    00002e0: ed19 07b6 00ed 1301 7db6 00ed 1d19 0511
    00002f0: 017a 0454 1906 1104 035c 2e04 604f b800
    0000300: f3b6 00ed 1303 8ab6 00ed b600 f7b7 00fa
    0000310: 1905 1101 7b04 54bf 1906 1104 055c 2e04
    0000320: 604f 1401 99b8 022c 1906 1104 075c 2e04
    0000330: 604f 1908 1905 1101 7c04 54b0          
  Exception Handler Table:
    bci [172, 517] => handler: 530
  Stackmap Table:
    append_frame(@85,Object[#1334],Object[#220])
    full_frame(@236,{Object[#2],Object[#19],Object[#367],Integer,Object[#22],Object[#1334],Object[#220],Object[#139],Null,Object[#797],Object[#966]},{})
    append_frame(@270,Integer)
    same_frame(@286)
    same_frame(@296)
    same_frame(@347)
    append_frame(@433,Object[#139])
    same_frame(@464)
    full_frame(@500,{Object[#2],Object[#19],Object[#367],Integer,Object[#22],Object[#1334],Object[#220],Object[#139],Object[#4],Object[#797],Object[#966],Integer,Object[#139]},{})
    full_frame(@530,{Object[#2],Object[#19],Object[#367],Integer,Object[#22],Object[#1334],Object[#220],Object[#139],Object[#4]},{Object[#274]})
    full_frame(@687,{Object[#2],Object[#19],Object[#367],Integer,Object[#22],Object[#1334],Object[#220],Object[#139],Null},{})
    full_frame(@792,{Object[#2],Object[#19],Object[#367],Integer,Object[#22],Object[#1334],Object[#220],Object[#139],Object[#4],Object[#797],Object[#966],Integer,Object[#139]},{})

	at android.content.res.Resources.__constructor__(Resources.java:219)
	at android.content.res.Resources.<init>(Resources.java)
	at android.content.pm.PackageParser.parseBaseApk(PackageParser.java:1259)
	at android.content.pm.PackageParser.parseMonolithicPackage(PackageParser.java:1212)
	at android.content.pm.PackageParser.parsePackage(PackageParser.java:969)
	at android.content.pm.PackageParser.parsePackage(PackageParser.java:981)
	at org.robolectric.shadows.ShadowPackageParser.callParsePackage(ShadowPackageParser.java:41)
	at org.robolectric.android.internal.AndroidEnvironment.loadAppPackage_measured(AndroidEnvironment.java:291)
	at org.robolectric.android.internal.AndroidEnvironment.lambda$loadAppPackage$1(AndroidEnvironment.java:261)
	at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:53)
	at org.robolectric.android.internal.AndroidEnvironment.loadAppPackage(AndroidEnvironment.java:259)
	at org.robolectric.android.internal.AndroidEnvironment.installAndCreateApplication(AndroidEnvironment.java:158)
	at org.robolectric.android.internal.AndroidEnvironment.setUpApplicationState(AndroidEnvironment.java:149)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:298)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:247)
	... 5 more

**Comment by Jagadeshmr** [comment]
I have been facing the same issue with chipmunk studio version, have anyone got the solution

**Comment by UlrichLa** [comment]
> I am using Android Studio Chipmunk Robo electric version 4.2.1
> 
> huh? java 11, need help! java.lang.RuntimeException: huh? java 11, need help! at 
> ...
> I had even upgraded roboelectric verion still same error , unit test with code coverage is failing

I have the same issue with the "huh? java 11, need help!" error and found a temporary solution:
Delete the "%USERPROFILE%\m2\repository\orgrobolectric\android-all\9-robolectric-4913185-2\android-all-9-robolectric-4913185-2.jar" file. After that it downloads it durring the testrun again and it works on myside, but only once. For the next testrun you must delete again.

P.S.: I'm new to the topic, so do not blame me if i did something completely wrong... ;-)

**Comment by hoisie** [comment]
I am still interested in finding a resolution to this, but I still have not been able to get a clean repro of the `VerifyError`. If anyone can put up a small sample app with repro instructions for the `VerifyError`, I'd be happy to take a look at it. If it's not reproducible any more it should be closed.

**Comment by hoisie** [comment]
OK I think I got a repro. It is using preinstrumented jars so it's not a problem with Robolectric's instrumentation. The AS code coverage tool is completely broken and throws the VerifyError. Jacoco at least doesn't crash but everything shows up as 0% coverage. At this point the best option may be to file a bug against IntelliJ and see if they can can take a look. I have no idea how the IntelliJ code coverage tool works under the hood.

**Comment by JorikFat** [comment]
I fix it!
![image](https://user-images.githubusercontent.com/23609498/197620550-7d0f6e43-7db3-4afc-a946-715f313f33d6.png)
need to include coverage classes and coverage now working!

**Comment by hfkai** [comment]
![image](https://user-images.githubusercontent.com/12794733/199413517-22beecc6-d650-4e8b-b5cb-ec8f13d5e9e2.png)
I have the same problem, can anyone help me? I have replaced the coverage runner with jacoco, but it still doesn't work at all. I also added -noverify to my vm option, and configured @Config(sdk = {Build.VERSION_CODES.P }) annotation, but still reported an error

**Comment by hfkai** [comment]
> I am still interested in finding a resolution to this, but I still have not been able to get a clean repro of the `VerifyError`. If anyone can put up a small sample app with repro instructions for the `VerifyError`, I'd be happy to take a look at it. If it's not reproducible any more it should be closed.

I meet this question，can you help me ？please

**Comment by funyin** [comment]
Finally fixed it. 
Just had to edit the run configuration and include the package in the coverage data like this

com.example.*

**Comment by sdlu1992** [comment]
> 

Your solution is effective, Thanks a lot!
also we can add exclude package so that dont need add our biz package one by one, just add this:

![image](https://user-images.githubusercontent.com/2403578/231125131-8d1b43e6-a469-470c-9141-22c3bd1a7b7f.png)

**Comment by zameelpichen** [comment]
@funyin that works!

my thesis is : its just confused of what to pick when multiple packages are there
![image](https://user-images.githubusercontent.com/25373198/232247102-9d467db7-77a6-4f5c-8508-84eacca18029.png)

so choosing one manually works
![image](https://user-images.githubusercontent.com/25373198/232247143-6950326f-2152-456d-98e7-b5ce0fb73c5d.png)

heres how to add the package for any future visitors!
![image](https://user-images.githubusercontent.com/25373198/232247223-6ea7b3cd-daf5-4f56-a675-a13c8fd4db1e.png)

**Comment by ignaciotcrespo** [comment]
> Finally fixed it. Just had to edit the run configuration and include the package in the coverage data like this
> 
> com.example.*

@funyin it works for me, thanks!
and thanks to @zameelpichen for the screenshots 👍

**Comment by yuliuslio** [comment]
> I fix it! ![image](https://user-images.githubusercontent.com/23609498/197620550-7d0f6e43-7db3-4afc-a946-715f313f33d6.png) need to include coverage classes or packages Now working!

This solution worked for me, thank you!

**Comment by hoisie** [comment]
Good catch everyone, the issue is that the built-in code coverage tool in Android studio crashes when analyzing the `android.*` classes that have been instrumented by Robolectric. It should work for everything else.

**Comment by ganadist** [comment]
Discussing at https://issuetracker.google.com/issues/319010149

**Comment by kozmotronik** [comment]
> I fix it! ![image](https://user-images.githubusercontent.com/23609498/197620550-7d0f6e43-7db3-4afc-a946-715f313f33d6.png) need to include coverage classes or packages Now working!

@JorikFat 's solution fixed the `java.lang.VerifyError: Bad return type` issues in my case. Thank you @JorikFat.

**Comment by cpele** [comment]
> turn on the -noverify switch for java.
This is not working for me.

I had that error when running my unit test from Android Studio:
```
Bad return type
Exception Details:
  Location:
    android/content/res/ResourcesImpl.$$robo$$android_content_res_ResourcesImpl$loadComplexColorForCookie(Landroid/content/res/Resources;Landroid/util/TypedValue;ILandroid/content/res/Resources$Theme;)Landroid/content/res/ComplexColor; @1000: areturn
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'android/content/res/ComplexColor' (from method signature)
```

What fixed it for me was to set the "specify classes and packages" option in the run configuration:

![image](https://github.com/user-attachments/assets/1980952c-2f60-453d-b050-8f96dfcfd0d9)

By adding my project's package which starts in `tv`

**Comment by hoisie** [comment]
Just FYI I filed an issue on JetBrains a while back for this:

https://youtrack.jetbrains.com/issue/IDEA-349117#focus=Comments-27-9796902.0-0

There hasn't been much activity on this issue yet.

**Comment by CaiJinFu** [comment]
Android Studio version：Koala Feature Drop | 2024.1.2
robolectric = "4.12.2"
powermock = "2.0.9"
In the build.gradle file of the app module, configure the following:
android {
    testOptions {
        unitTests {
            includeAndroidResources = true
            all {
                // Enabling this configuration activates the coverage
                jvmArgs '-noverify'
            }
        }
    }
}

**Comment by simplekjl** [comment]
nice, how to setup this in the CI?

**Comment by kuraki-z** [comment]
@funyin @zameelpichen Thank you!It works,This has been a problem that has troubled me for months.

**Comment by Abenezer-W** [comment]
I didn't want to go in and change the configuration for each test and i also wanted to share what worked for me because some of the solutions above have older android studio ui's. Converting those UI actions to the most recent android studio version can be a pain so here is my solution

add this to you build.gradle.kts
```
testOptions {
    unitTests {
        isIncludeAndroidResources = true
        all {
            it.jvmArgs("-noverify")
        }
    }
}

```

setups 
sdk: 36
Android Studio Narwhal 3 Feature Drop | 2025.1.3

**Comment by kenyee** [comment]
FWIW, using -noverify didn't work for me.. AS kept showing 100% coverage on any test suite that used robolectric 😂 
Editing the run config to exclude android/androidx is probably the right thing to do, but it also doesn't work 🤔 (it keeps showing coverage for the first set of tests I tried running coverage on before adding these configurations so everything is 100% coverage).
This is on latest Panda/Otter.

**Comment by hoisie** [comment]
Hey everyone, I think I made a breakthrough on this issue. I assumed it was an Android Studio coverage issue, but it is likely a Robolectric issue.

I spent like 5 hours debugging this with Gemini. Thankfully the IntelliJ coverage agent is open-source: https://github.com/JetBrains/intellij-coverage.

I built the agent locally, and modified the agent to print the classes before and after the coverage instrumentation. Then using Gemini, I compared the bytecode of the problematic method (ResourcesImpl.loadComplexColorForCookie) both before and after instrumentation. Gemini said that the Stack Map Table for the instrumented version was not correct, and that it assigned the common superclass of `GradientColor` and `ColorStateList` a `Object`, when it should be `ComplexColor`.

Then, looking at the agent's [ClassWriterImpl.getCommonSuperClass](https://github.com/JetBrains/intellij-coverage/blob/master/util/src/com/intellij/rt/coverage/instrumentation/ClassWriterImpl.java#L46), it loads classes using [ClassLoader.getResourceAsStream](https://github.com/JetBrains/intellij-coverage/blob/master/util/src/com/intellij/rt/coverage/instrumentation/ClassWriterImpl.java#L140). This is where I began to suspect that it could be a Robolectric issue.

I added some dumping logic to ClassWriterImpl when it computed the common superclass, and noticed that loading ColorStateList returned the stubs jar byte code!

I realized that it's possible that when you try to load a class in Robolectric as a resource, it will not load the local URL from the android-all jar, but the one from the stubs jar.

I wrote a quick test to verify it and it was indeed the issue, and I tried coverage with that and it worked. I put up #10980 to fix the issue.

**Comment by DevDengChao** [comment]
@hoisie Hope it will be merged soon, hero.

**Comment by hoisie** [comment]
I will push out a minor release with this fix and other fixes.

**Comment by Marko2407** [comment]
Hi, when we can expect this to be released 🤞🏻
